File: COMP4958/LectureNotes/08/shape.go
package main

import (
	"fmt"
	"math"
)
type Shape interface {
	Area() float64
}

type Circle struct {
	x, y, r float64
}

func (c Circle) Area() float64 {
	return c.r * c.r * math.Pi
}

type Rectangle struct {
	x1, y1, x2, y2 float64
}
func (r Rectangle) Area() float64 {
	return math.Abs(r.x2 - r.x1) * math.Abs(r.y2 - r.y1)
}

func totalArea(shapes []Shape) float64 {
	total := 0.0
	for _, s:= range shapes {
		total += s.Area()
	}
	return total 
}

func main() {
	shape := []Shape {
		Circle{0.0, 0.0, 1.0},
		Rectangle{1.0, 2.0, 3.0, 4.0},
		Circle{0.0, 0.0, 2.0},
	}
	fmt.Println(totalArea(shape))
}

File: COMP4959/Lectures/04/shape.rs
trait Shape {
    fn area(&self) -> f32;

}

struct Circle (f32, f32, f32);

impl Shape for Circle {
  fn area (&self) -> f32 {
    std::f32::consts::PI * self.2 * self.2
  }
}

struct Square (f32);

impl Shape for Square {
  fn area (&self) -> f32 {
    self.0 * self.0
  }
}

fn total_area(shapes: &Vec<Box<dyn Shape>>)->f32 {
  let mut total = 0.0;
  for shape in shapes {
    total += shape.area();
  }
  total
}

fn main() {
  let v:Vec<Box<dyn Shape>> = vec![Box::new(Circle(0.0, 0.0, 1.0)),
              Box::new(Square(1.0))];
  println!("Total area: {}", total_area(&v));
}